File: windows/src/TraceCommons.Interop/QueueGrouping.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace TraceCommons.Interop;

/// <summary>
/// One project's slice of the queue: every pending entry that shares a
/// <c>project_id</c>, and what the "Submit all" group action needs to know
/// about them.
/// </summary>
/// <remarks>
/// Deliberately holds counts, not entries. The queue row list already owns
/// the entries themselves; a group is a view over the same data, not a
/// second copy of it, so this carries only what a folder row renders and what
/// a project-group submit needs to send.
///
/// Under the folder-first layout a row renders two more scalars -- the
/// project's display path and the total bytes waiting in it -- and they are
/// here for exactly the same reason the count is: they are facts about the
/// group, computed once in the grouping pass, not a second copy of the
/// entries.
/// </remarks>
public sealed class ProjectQueueGroup
{
    internal ProjectQueueGroup(
        string projectId,
        string projectLabel,
        string projectPath,
        int count,
        long sizeBytes)
    {
        ProjectId = projectId;
        ProjectLabel = projectLabel;
        ProjectPath = projectPath;
        Count = count;
        SizeBytes = sizeBytes;
    }

    /// <summary>
    /// The id every entry in this group shares -- the one <c>entry_value</c>
    /// publishes as <c>project_id</c>, and the one a group's "Submit all"
    /// must send. Never <see cref="ProjectLabel"/>, which is display text and
    /// not stable: two entries can carry the same label under different ids,
    /// or the same id under a label that changed, and grouping on the label
    /// would silently merge or split groups the daemon does not consider the
    /// same project.
    /// </summary>
    public string ProjectId { get; }

    /// <summary>The display label shown on the folder row.</summary>
    public string ProjectLabel { get; }

    /// <summary>
    /// The project's folder, <c>~</c>-abbreviated, taken from the first entry
    /// in the group. Empty against a daemon predating the field, in which
    /// case the row renders its label alone.
    /// </summary>
    /// <remarks>
    /// Every entry sharing a <see cref="ProjectId"/> shares a project key, so
    /// the first member's path is the group's path -- there is no
    /// disagreement to reconcile. It is display-only: renderable, never
    /// logged, audited, notified, or persisted to history. See
    /// <see cref="QueueEntry.ProjectPath"/>.
    /// </remarks>
    public string ProjectPath { get; }

    /// <summary>How many pending entries this project has.</summary>
    public int Count { get; }

    /// <summary>
    /// The session bytes on disk this folder is holding, summed over its
    /// entries.
    /// </summary>
    /// <remarks>
    /// Sessions on disk, never a would-send figure. That number only a
    /// preview computes, and stating one as the other on a consent surface
    /// is the app's first false statement about what leaves the machine --
    /// the same rule the card's own size line follows.
    /// </remarks>
    public long SizeBytes { get; }

    /// <summary>
    /// Whether the folder row offers a "Submit all" action. Always -- see the
    /// remark.
    /// </summary>
    /// <remarks>
    /// This used to be <c>Count &gt; 1</c>, on the reasoning that a
    /// single-entry group's own row already had a Submit doing exactly the
    /// same thing. That was true of a flat list where the row and the header
    /// were on screen together. Under the folder-first layout the row is a
    /// level down, so hiding this would mean opening a folder to do the thing
    /// the folder is offering. The rule expired with the layout it was
    /// written for; the property stays so callers do not have to know that.
    /// </remarks>
    public bool ShowSubmitAll => true;
}

/// <summary>
/// Buckets the pending queue by project, for the group headers and their
/// <c>Submit all</c> action.
///
/// This is pure logic -- no UI type, no I/O -- so it is testable here, on a
/// machine that cannot build WinUI at all. The bucketing rule (project_id,
/// never project_label), the group order, and whether a group's action
/// appears are exactly the decisions a rendering bug could get wrong
/// silently; keeping them here means a wrong bucket or a wrongly-shown
/// action is a red test today rather than a screenshot review someday.
/// </summary>
public static class QueueGrouping
{
    /// <summary>
    /// Groups <paramref name="entries"/> by <see cref="QueueEntry.ProjectId"/>.
    ///
    /// Group order is first-seen: the order the first entry of each project
    /// appears in <paramref name="entries"/>, which is also the order the
    /// queue itself is in. A contributor who has already scanned the list
    /// top to bottom should not see it reshuffle the moment it grows a
    /// header.
    ///
    /// An entry with no project id (<c>null</c> or empty) groups with every
    /// other entry that also has none, under an empty id -- it does not
    /// silently disappear from the queue, and it does not get a project id
    /// invented for it.
    /// </summary>
    public static IReadOnlyList<ProjectQueueGroup> ByProject(IEnumerable<QueueEntry> entries)
    {
        ArgumentNullException.ThrowIfNull(entries);

        var order = new List<string>();
        var counts = new Dictionary<string, int>(StringComparer.Ordinal);
        var labels = new Dictionary<string, string>(StringComparer.Ordinal);
        var paths = new Dictionary<string, string>(StringComparer.Ordinal);
        var bytes = new Dictionary<string, long>(StringComparer.Ordinal);

        foreach (QueueEntry entry in entries)
        {
            ArgumentNullException.ThrowIfNull(entry);
            string key = KeyOf(entry);

            if (!counts.ContainsKey(key))
            {
                order.Add(key);
                counts[key] = 0;
                labels[key] = LabelOf(entry);
                paths[key] = entry.ProjectPath;
                bytes[key] = 0;
            }

            counts[key]++;
            bytes[key] += entry.SizeBytes;
        }

        return order
            .Select(key => new ProjectQueueGroup(
                key,
                labels[key],
                paths[key],
                counts[key],
                bytes[key]))
            .ToList();
    }

    /// <summary>
    /// The bucketing key for one entry -- <see cref="QueueEntry.ProjectId"/>,
    /// empty-string if absent, never <see cref="QueueEntry.ProjectLabel"/>.
    /// Public so a caller reconstructing which rows belong to a
    /// <see cref="ProjectQueueGroup"/> (this method reports counts, not
    /// membership -- see <see cref="ProjectQueueGroup"/>'s remarks) uses
    /// exactly the same rule <see cref="ByProject"/> bucketed them with,
    /// rather than restating it and risking the two rules drifting apart.
    /// </summary>
    public static string KeyOf(QueueEntry entry)
    {
        ArgumentNullException.ThrowIfNull(entry);
        return entry.ProjectId ?? string.Empty;
    }

    /// <summary>
    /// The label a group's header shows: the project's own label, falling
    /// back to the id, falling back to a fixed placeholder. Never a path,
    /// which the daemon does not send and this must not invent. Matches
    /// <c>QueueEntryViewModel.ProjectLabel</c>'s fallback chain exactly --
    /// a row and its own group must never disagree about what to call the
    /// project it is in.
    /// </summary>
    private static string LabelOf(QueueEntry entry) =>
        !string.IsNullOrWhiteSpace(entry.ProjectLabel) ? entry.ProjectLabel!
        : !string.IsNullOrWhiteSpace(entry.ProjectId) ? entry.ProjectId!
        : "Unknown project";
}

File: macos/Sources/TCShellCore/QueueGrouping.swift
import Foundation

/// One project's slice of the waiting queue: the entries themselves, plus
/// the totals the group header states.
///
/// The entries live *in* the group rather than being filtered back out of
/// the flat list by the view. That is the whole reason this type exists --
/// see `QueueGrouping`.
public struct QueueGroup<Entry>: Identifiable {
    /// `project_id`, and the id `submitProject` takes. Never the label: a
    /// label is a display name, is not guaranteed unique across two
    /// different projects, and grouping on it would merge them into one
    /// bucket with one `Submit all` that could approve the wrong project's
    /// entries.
    public let id: String
    /// The label the first entry of this project carried. A later entry
    /// with a stale label does not rename the group out from under its
    /// buttons.
    public let label: String
    /// Sum of the group's entry sizes.
    public let bytes: Int
    public let entries: [Entry]

    public var count: Int { entries.count }

    public init(id: String, label: String, bytes: Int, entries: [Entry]) {
        self.id = id
        self.label = label
        self.bytes = bytes
        self.entries = entries
    }
}

extension QueueGroup: Equatable where Entry: Equatable {}

/// Groups the waiting queue by project in a single pass.
///
/// This replaces a shape that cost entries times projects on *every*
/// SwiftUI body evaluation: the view grouped the waiting list to get the
/// project headers, and then, once per header, filtered the whole waiting
/// list again to find that project's rows. At the 500-entry cap the queue
/// actually runs at, with a dozen projects, that was tens of thousands of
/// entry visits and a dozen freshly allocated arrays per redraw -- and a
/// fresh array every time also denies SwiftUI any chance of deciding a
/// group had not changed.
///
/// Grouping once, off the model rather than out of the view, makes that
/// cost proportional to what actually moved.
public enum QueueGrouping {
    public static func groups<Entry>(
        _ entries: [Entry],
        projectID: (Entry) -> String,
        projectLabel: (Entry) -> String,
        sizeBytes: (Entry) -> Int
    ) -> [QueueGroup<Entry>] {
        var order: [String] = []
        var labels: [String: String] = [:]
        var bytes: [String: Int] = [:]
        var members: [String: [Entry]] = [:]

        for entry in entries {
            let id = projectID(entry)
            if members[id] == nil {
                order.append(id)
                labels[id] = projectLabel(entry)
                bytes[id] = 0
                members[id] = []
            }
            bytes[id]! += sizeBytes(entry)
            members[id]!.append(entry)
        }

        return order.map { id in
            QueueGroup(
                id: id,
                label: labels[id] ?? "",
                bytes: bytes[id] ?? 0,
                entries: members[id] ?? []
            )
        }
    }
}
